src/Shared/Targets/Wrappers/AspNetBufferingTargetWrapper.cs

```diff
@@ -0,0 +1,301 @@
+        {
+        
+            WrappedTarget      = wrappedTarget;
+            BufferSize         = bufferSize;
+            GrowBufferAsNeeded = true;
+        }
+
+        /// <summary>
+        /// Gets or sets the number of log events to be buffered.
+        /// </summary>
+        /// <docgen category='Buffering Options' order='100' />
+        [DefaultValue(100)]
+        public int BufferSize { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether buffer should grow as needed.
+        /// </summary>
+        /// <value>A value of <c>true</c> if buffer should grow as needed; otherwise, <c>false</c>.</value>
+        /// <remarks>
+        /// Value of <c>true</c> causes the buffer to expand until <see cref="BufferGrowLimit"/> is hit,
+        /// <c>false</c> causes the buffer to never expand and lose the earliest entries in case of overflow.
+        /// </remarks>
+        /// <docgen category='Buffering Options' order='100' />
+        [DefaultValue(false)]
+        public bool GrowBufferAsNeeded { get; set; }
+
+        /// <summary>
+        /// Limits the amount of slots that the buffer should grow
+        /// </summary>
+        private int _bufferGrowLimit;
+
+        /// <summary>
+        /// Gets or sets the maximum number of log events that the buffer can keep.
+        /// </summary>
+        /// <docgen category='Buffering Options' order='100' />
+        public int BufferGrowLimit
+        {
+            get => _bufferGrowLimit;
+
+            set
+            {
+                _bufferGrowLimit = value;
+                GrowBufferAsNeeded = (value >= BufferSize);
+            }
+        }
+
+        /// <summary>
+        /// Context for DI
+        /// </summary>
+        private IHttpContextAccessor _httpContextAccessor;
+
+        /// <summary>
+        /// Provides access to the current request HttpContext.
+        /// </summary>
+        /// <returns>HttpContextAccessor or <c>null</c></returns>
+        internal IHttpContextAccessor HttpContextAccessor
+        {
+            get => _httpContextAccessor ?? (_httpContextAccessor = RetrieveHttpContextAccessor());
+            set => _httpContextAccessor = value;
+        }
+
+        private IHttpContextAccessor RetrieveHttpContextAccessor()
+        {
+#if ASP_NET_CORE
+            return ServiceLocator.ResolveService<IHttpContextAccessor>(ResolveService<IServiceProvider>(), LoggingConfiguration);
+#else
+            return new DefaultHttpContextAccessor();
+#endif
+        }
+
+        /// <summary>
+        /// The Key for the buffer dictionary in the HttpContext.Items collection
+        /// </summary>
+        private static readonly object HttpContextItemsKey = new object();
+
+        /// <summary>
+        /// The lock obtained before the dictionary is created or the slot
+        /// in the dictionary is created.
+        /// </summary>
+        private readonly object _lock = new object();
```

Comment:
Should this be static?  What if the NLog.Config has several BufferingTargetWrappers?

Reply:
In the "simple" world, then one could could get rid of the dictionary, and just have each target-wrapper to have their own key. In the "real" world, then HttpContext-Items is actually not threadsafe, and the correct thing would be that the middleware upfront created a dictionary and inserted into the HttpContext-Items (using a static key). And then target-wrappers would use the same static-key to insert themselves in the already created dictionary.

Reply:
Ok that makes sense, I will modify that.

Reply:
Completed, HttpModule and Middleware calls Buffering.TargetWrapper.Initialize(HttpContext) at startup phase.

Reply:
Think `_lock` shoudl be removed, and instead once should perform the synchronization around the pre-allocated dictionary.

Reply:
Completed, did not catch that as a remnant from the previous method used.